File: backend/features/inquiries/inquiries-core/src/main/java/de/bennyboer/kicherkrabbe/inquiries/Sender.java
package de.bennyboer.kicherkrabbe.inquiries;

import jakarta.annotation.Nullable;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Value;
import lombok.With;

import java.util.Optional;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.NONE;
import static lombok.AccessLevel.PRIVATE;

@Value
@With(PRIVATE)
@AllArgsConstructor(access = PRIVATE)
public class Sender {

    SenderName name;

    EMail mail;

    @Nullable
    @Getter(NONE)
    PhoneNumber phone;

    public static Sender of(SenderName name, EMail mail, @Nullable PhoneNumber phone) {
        notNull(name, "Sender name must be given");
        notNull(mail, "Mail must be given");

        return new Sender(name, mail, phone);
    }

    public Optional<PhoneNumber> getPhone() {
        return Optional.ofNullable(phone);
    }

    public Sender anonymize() {
        return withName(getName().anonymize())
                .withMail(getMail().anonymize())
                .withPhone(null);
    }

    @Override
    public String toString() {
        return "Sender(name=%s, mail=%s, phone=%s)".formatted(name, mail, phone);
    }

}

File: frontend/management/src/app/modules/admin/modules/mailbox/model/sender.ts
import { Option, someOrNone, validateProps } from '@kicherkrabbe/shared';

export class Sender {
  readonly name: string;
  readonly mail: string;
  readonly phone: Option<string>;

  private constructor(props: { name: string; mail: string; phone: Option<string> }) {
    validateProps(props);

    this.name = props.name;
    this.mail = props.mail;
    this.phone = props.phone;
  }

  static of(props: { name: string; mail: string; phone?: string | null }): Sender {
    return new Sender({
      name: props.name,
      mail: props.mail,
      phone: someOrNone(props.phone),
    });
  }
}
